### DISCUSSION TITLE
Chord Rendering

### DISCUSSION THREAD

**Discussion by thecodeinventor** [no-chris/chord-mark#660] [Ideas]
This is an awesome library! I started using it a while ago and it's just a pleasure to use. Great work! ... and please, keep up the good work! :)

I'm planning on using this library a lot so, if it's helpful, I can provide some user feedback. Here are some things that came up so far:

1. I was wondering if there is some possibility to define how some chords should be parsed?
ex: "CMA7" will be rendered as "CM7" but I would like it to be "Cmaj7"  (it's about familiar visual cues) 

2. Repeating chords will cause the ChordLine not to render if chords are subdivision of a measure.
ex: these will not render: 
- A. A. C. NC.
- [A A] E. F. A.
- [NC NC A] E. F. A.

3. The following example will not render the square brackets next to "D B"
2/4
A. [E C] [D B] G. 

4. *I noticed that in "key Gm" F# will be rendered as Gb. Is there for the user to control this? (asking for a friend)

Thank you!

**Comment by no-chris** [comment]
Hello!

Thanks for the comments 👍🏻 

1. absolutely, you can define a custom chord renderer based on the chord-symbol library (https://github.com/no-chris/chord-symbol). You can take inspiration from here: https://github.com/no-chris/chord-mark/tree/master/packages/chord-mark-converters#to-ultimate-guitar-format

2. This is by design, is there any specific reason you prefer to write 
A. A. C. NC.
rather than
A.. C. NC.
?

3. duh, that's a nice bug! Will try to look into it asap

4. Not at the moment, the choice of flat/sharp is based on the key itself, and as Gm is using flats then the chords are normalized as flats. You can force the rendering to use sharp using the `accidentalsType` option if that works for you (https://github.com/no-chris/chord-mark/blob/347939373e55503f6c6c7ee0d3f13e7bf18940f6/packages/chord-mark/src/renderer/components/renderSong.js#L28)
I'm thinking anyway that there should be a way to opt out chord normalization as it might be too opinionated to be useful for everyone.

**Comment by thecodeinventor** [comment]
1. This was very helpful, thank you. I'm using "chordSymbolRenderer: chordRendererFactory(options)" and by default this is causing some unpleasant renderings: m->mi, sus2->(add9,omit3). Do I need to handle these manually in a customFilter or can this be done easier? If manually, is there a full list of how the chords will be rendered so I can handle them in a customFilter?
2. There were some instances where such a notation (repeating a chord or NC) could have suggested rhythm better. 
3. Thank you!
4. Opting out of chord normalization would keep the #/b of the note regardless of the key? This could be a simple solution especially for keys where we can encounter the use of both types of accidentals.

**Comment by no-chris** [comment]
1. you are probably looking for the `useShortNamings` option
https://github.com/no-chris/chord-symbol/blob/master/packages/chord-symbol/API.md#rendererconfiguration--object
https://github.com/no-chris/chord-symbol/tree/master/packages/chord-symbol#rendering-and-normalization

2. interesting, I'll keep that in mind 👍🏻 
3.
4. And it would also solve issue number 1. Cannot commit to any timeframe but this is something I had in mind for a long time

**Comment by thecodeinventor** [comment]
1. yes, that was it.  
Note: when using the `chordRendererFactory` and forcing accidentals type, it's important to know that chord-mark uses `accidentalsType = ('auto' | 'flat' | 'sharp)'` and chord-symbol uses `accidental : ('original' | 'flat' | 'sharp')`

so far, so good!

**Comment by no-chris** [comment]
3. fixed in https://github.com/no-chris/chord-mark/pull/663

**Comment by thecodeinventor** [comment]
I am successfully rendering a chart with all the options, including using a `chordSymbolRenderer = chordSymbol.chordRendererFactory()`. Works great: transposing, setting accidental type and even using customFilters. 
The issue is that in this configuration automatic transposition fails and it seems to do so only while using a `chordSymbolRenderer` . 

ex: 
key D
#v
Dm7 G7 Cmaj7 %
The first verse is in the key of C

key G
#v
And the second one in the key of G!

In this case, the second key declaration `<span class="cmKeyDeclaration">key: G</span>` is rendered inside the `<div class="cmSection cmSection-Verse">` of the first key and is basically in line with the lyrics of Verse 1. The automatic transposition between the two verses does not work (both verses have same chords) but general transposition for the song works (everything else works fine too: customFilters, etc.). If not using a `chordSymbolRenderer`, the automatic transposition of verse 2 works fine but the rendering is still the same as above.

Am I not handling it correctly or bug?

**Reply by no-chris** [reply]
Hello
Can you please share a code sample?
Thanks

**Reply by thecodeinventor** [reply]
Sure! 

```
function CustomChords() {
    return function (chord) {
        chord.formatted.symbol = chord.formatted.symbol
            .replace('M7', 'maj7')
            .replace('M9', 'maj9');
        return chord;
    };
};
```

```
function generateRenderOptions(transposeValue, accidentalsType){
    let options = {};
    let rendererFactoryOptions = { // init
        customFilters: [
            CustomChords()
        ],       
        useShortNamings: true,
        simplify : 'none', 
        transposeValue : 0, 
        accidental : 'original',
        symbolType : 'chord',
        printer : "text"
    };

    // configure rendererFactory options
    rendererFactoryOptions.transposeValue = transposeValue;
    rendererFactoryOptions.accidental = accidentalsType == 'auto' ? 'original' : accidentalsType;
    
    // define chordSymbol renderer
    options.chordSymbolRenderer = chordSymbol.chordRendererFactory(rendererFactoryOptions);
    
    // set general options 
    options.transposeValue = transposeValue;
    options.accidentalsType = accidentalsType ? accidentalsType : "auto";
   
    return options
}
```

```
function renderSongChart(songChartCode, options){    
    const parsed = chordMark.parseSong(songChartCode);    
    const rendered = chordMark.renderSong(parsed, options);  
    return rendered;
}
```

calling `renderSongChart(songChartCode, generateRenderOptions(0, "auto"));` produces the following rendering of the chord-mark code from my previous post: (rendering issue + chords are not automatically transposed between the two verses)
```
<div class="cmSong">
   <p class="cmLine"><span class="cmKeyDeclaration">key: D</span></p>
   <div class="cmSection cmSection-Verse">
      <p class="cmLine"><span class="cmSectionLabel">Verse 1</span></p>
      <p class="cmLine"><span class="cmChordLine"><span class="cmBarSeparator">|</span><span class="cmChordSymbol">Dm7</span>     <span class="cmBarSeparator">|</span><span class="cmChordSymbol">G7</span>     <span class="cmBarSeparator">|</span><span class="cmChordSymbol">Cmaj7</span>     <span class="cmBarSeparator">|</span><span class="cmChordSymbol">%</span>     <span class="cmBarSeparator">|</span></span></p>
      <p class="cmLine"><span class="cmLyricLine">The first verse is in the key of C</span></p>
      <p class="cmLine"><span class="cmEmptyLine">&nbsp;</span></p>
      <p class="cmLine"><span class="cmKeyDeclaration">key: G</span></p>    <!-- rendering issue?: should be outside of <div> -->
   </div>
   <div class="cmSection cmSection-Verse">
      <p class="cmLine"><span class="cmSectionLabel">Verse 2</span></p>
      <p class="cmLine cmLine--isFromAutoRepeatChords"><span class="cmChordLine"><span class="cmBarSeparator">|</span><span class="cmChordSymbol">Dm7</span>     <span class="cmBarSeparator">|</span><span class="cmChordSymbol">G7</span>     <span class="cmBarSeparator">|</span><span class="cmChordSymbol">Cmaj7</span>     <span class="cmBarSeparator">|</span><span class="cmChordSymbol">%</span>     <span class="cmBarSeparator">|</span></span></p>
      <p class="cmLine"><span class="cmLyricLine">And the second one in the key of G!</span></p>
   </div>
</div>
```
if we comment out the line: `// options.chordSymbolRenderer = chordSymbol.chordRendererFactory(rendererFactoryOptions);` then this is the rendered result: (same rendering issue, but chords are transposed between the two verses)
```

<div class="cmSong">
   <p class="cmLine"><span class="cmKeyDeclaration">key: D</span></p>
   <div class="cmSection cmSection-Verse">
      <p class="cmLine"><span class="cmSectionLabel">Verse 1</span></p>
      <p class="cmLine"><span class="cmChordLine"><span class="cmBarSeparator">|</span><span class="cmChordSymbol">Dm7</span>     <span class="cmBarSeparator">|</span><span class="cmChordSymbol">G7</span>     <span class="cmBarSeparator">|</span><span class="cmChordSymbol">CM7</span>     <span class="cmBarSeparator">|</span><span class="cmChordSymbol">%</span>     <span class="cmBarSeparator">|</span></span></p>
      <p class="cmLine"><span class="cmLyricLine">The first verse is in the key of C</span></p>
      <p class="cmLine"><span class="cmEmptyLine">&nbsp;</span></p>
      <p class="cmLine"><span class="cmKeyDeclaration">key: G</span></p>    <!-- rendering issue? -->
   </div>
   <div class="cmSection cmSection-Verse">
      <p class="cmLine"><span class="cmSectionLabel">Verse 2</span></p>
      <p class="cmLine cmLine--isFromAutoRepeatChords"><span class="cmChordLine"><span class="cmBarSeparator">|</span><span class="cmChordSymbol">Gm7</span>     <span class="cmBarSeparator">|</span><span class="cmChordSymbol">C7</span>     <span class="cmBarSeparator">|</span><span class="cmChordSymbol">FM7</span>     <span class="cmBarSeparator">|</span><span class="cmChordSymbol">%</span>     <span class="cmBarSeparator">|</span></span></p>
      <p class="cmLine"><span class="cmLyricLine">And the second one in the key of G!</span></p>
   </div>
</div>

```

**Comment by thecodeinventor** [comment]
Is it that bad?

**Reply by no-chris** [reply]
Sorry for taking too long to respond

There is a fundamental flaw in the logic here:
https://github.com/no-chris/chord-mark/blob/d2ed1c3d49472fc2635bcf14f9d05baf09e19bd1/packages/chord-mark/src/renderer/helpers/renderAllChords.js#L76

If a custom renderer is given, then it is returned as it is
if not, then a renderer is created and only this one gets the proper transpose/accidentals parameters that are set later in the function

This is bad indeed, I need to think this through in the next few days as I don't see a quick and easy fix

I will create an issue to track this, many thanks for finding the issue and bringing it up 🙌🏻

**Reply by thecodeinventor** [reply]
It's my pleasure to work with this library and provide feedback. Thank you for maintaining it!

Did you look into the rendering issue with the second key declaration too? 
`<p class="cmLine"><span class="cmKeyDeclaration">key: G</span></p>    <!-- rendering issue?: should be outside of <div> -->`

**Reply by no-chris** [reply]
This is not a rendering issue per se, basically anything that is not a section declaration goes into the section's div
I agree it's not the most elegant from a markup point of view but I think it works well enough like that, unless it prevents you to achieve some use case?

**Reply by thecodeinventor** [reply]
The only concern about this is styling. By default, the first key declaration looks different than the following key declarations (ex: the first key declaration is not indented but the verses do have an indent which is inherited by the second key declaration). It will just take some extra CSS to style. No problem...
I only mentioned this aspect because I compared my rendering visually to the "chord-chart-studio" rendering and thought it would be of some use to point this out.

**Comment by no-chris** [comment]
I'm thinking of 2 approaches to fix the issues:

1. first is to change (or extend) the chord-symbol API to allow changing the renderer parameters dynamically. But that would be kind of moving away from the functional approach to maybe a more object oriented model with a renderer maintaining a mutable state. Not a big fan, but that's a possibility

2. the other approach is more radical and would get rid of custom filters completely in Chord Symbol. Because whichever logic you put in there can be probably be achieved by wrapping the current renderer in a custom function that would post process the output of the Chord Symbol renderer. And those custom filters may seems like a nice idea but they create a lot of dependencies to the chord-symbol internals as they expose too much. REgarding ChordMark, it could take as parameter a chordPostProcessor function instead of the chordSymbolRenderer that would allow achieving the same results

Leaning more towards 2 but will give it a few more days to think about it... if you have any comments/ideas, feel free to share as well

**Reply by thecodeinventor** [reply]
If I understand correctly, then what you are describing sounds pretty exciting because:

1. If this would allow access to the Chord Symbol Chord as an object right before returning it as a text, one could:
- make changes to the chord (ex: M7 -> maj7, M9 -> maj9, etc)
- insert HTML in the chord for extra styling (ex: Cmaj7/E -> wrap **C** in a `<span class="chord-root">`, wrap **maj7** in a `<span class="chord-color">`, wrap **/** in a `<span class="slash">`, wrap **E** in a `<span class="chord-bass">`). Using those clases, the chords could look more stylish and easier to read
- implement exceptions/corrections (ex: in key Gm, Gb -> F#, etc)
- use special fonts (ex: there is a font somewhere that uses special chars to render chords: something like "[" is "#", "]" is "b", but if the chord also has a "# 5" then that "#" would be "^" and would be rendered as superscript. The end result is a cool rendering of the chord. And yes, this would probably mess up spacing, but when rendering only chords and bars, it would look awesome)
- generate a list of all the unique chords in the song and associate them with instrument renderings of those chords. (ex: guitar shapes, piano shapes, etc.)

All this would be possible because Chord Symbol isolates each component of a Chord.

2. This is functional: it will allow for simple post processing of the final text form of the Chord. I guess _some_ of the above could also be also done here, but not as easy as above.

Is this wishful thinking or pretty doable? 😁

**Reply by no-chris** [reply]
Very interesting!
I think we can make all this work regardless of the approach but yes, it might make more sense to maintain custom filters and to use the raw object in ChordMark. Will try to dedicate some cycles to this later this month

**Comment by thecodeinventor** [comment]
Rendering side quests: (inspired by "...if you have any comments/ideas, feel free to share as well" )

1. Pickup measures. A way to define a measure that is an exception to the time signature. These need to be distinguishable so that the player will not count a full measure. 

Notes: 
- in this measure, the beat count does not have to add up to the time signature - one just has to mark somehow the measure starts and where it ends.
- if no "_" are present, this pickup measure should render as short as possible every time, even if there is enough space available 
- the pickup measure should not align with other measures, or it should align on the outside) 
```
    | C..                         F.   G.   |C             |
|G  | C..                         F.   G.   |C             |
```

2. Would it be difficult to implement some rule to allow measures to be broken up between chord rows? 
Sure, this could be resolved by moving the lyrics in the same line, but that would mess up the lyrics layout (phrasing) for the singer. 

ex: in following sequence the "G" chord needs to be on the next chord line (on "Con"), but it won't render because a measure needs to add up to 4 beats in the same row.
```
|A                  |D...    G.  | 
_ Lorem ipsum dolor _sit amet, 
     |A             |
_Con _sectetur adipiscing elit
```

Expected result: (this rendering should be possible if the source code would contain an "\\" (or some other character) after the "D..." chord)
```
|A                  |D...  
_ Lorem ipsum dolor _sit amet, 
G.   |A             |
_Con _sectetur adipiscing elit
```

Please let me know what you think of these ideas. We can discuss them in more detail if necessary... or not. I don't want to impose.
Anyway, these ideas (and others) will become more clear and useful the more I use this library.

**Reply by no-chris** [reply]
Thanks for all those suggestions!

1. Sorry to ask but what is a pickup measure, can you give an example in a well known song?

2. Yes, this has bothered me as well in the past for many transcriptions. I'm thinking that this could be a rendering thing, e.g. the writing of the chord chart could be with the lyrics layout "broken" but the "correct" bar structure, with a symbol denoting where the new line should actually start, eg:
```
A D...    G. 
_ Lorem ipsum dolor _sit amet, \_Con
A
_sectetur adipiscing elit
``` 

Here I'm using the backslash, and the rendering could be as you suggested or almost:
```
|A                |D...  
Lorem ipsum dolor sit amet, 
G.  |A                       |
Con sectetur adipiscing elit
```

Or it could only apply to the "lyrics only" rendering, if the target user is really the singer
```
Lorem ipsum dolor sit amet, 
Con sectetur adipiscing elit
```

Actually this is very close to what you are suggesting

**Comment by thecodeinventor** [comment]
1.
> A pickup measure (formal name: anacrusis) is a partial measure of notes that precedes the downbeat (strong beat) of the first, full measure. It is sometimes also called an “incomplete measure” or an “upbeat”. The notes of the pickup measure are referred to as “pickup notes”. Because a pickup measure is “incomplete”, it must always be completed at the end of a music piece/section by another partial measure, so that the result is one complete measure. [Source](https://www.dacapomusic.ca/blog/pickup-measures-explained)

ex: Amazing Grace, Happy Birthday, What A Wonderful World, etc.

Pickup measures are used when composers do not want to start a piece of music on a strong (down) beat. Pickup measures can contain only melody notes (No Chord) or can have Chords that either build up to the first down beat of the song, or create some kind of tension (ex: a 7 Chord). Anyway, right now, in a 4/4 song, there is no way to render a measure shorter than 4 beats.

2. Yeah, that is exactly what I was saying! Of course, I suppose your approach might be easier to implement technically (because you understand the existing code). I would write out the lyrics in their final form, and then use the "\" symbol to mark on the Chord line where a measure should be split up (seems more natural to me this way).

```
|A                  |D...    \ 
_ Lorem ipsum dolor _sit amet,
G.   |A             |
_Con _sectetur adipiscing elit
```

 In the end, the same result is accomplished. The only remaining "mystery" is how the measure bars will align? (ex: when displaying only chords)
```
|A                  |D...    G.   |
|A                  |
```
or like this:
```
|A                  |D...   
 G.   |A            |
```

**Reply by no-chris** [reply]
> ex: Amazing Grace, Happy Birthday, What A Wonderful World, etc.
> Pickup measures are used when composers do not want to start a piece of music on a strong (down) beat. Pickup measures can contain only melody notes (No Chord) or can have Chords that either build up to the first down beat of the song, or create some kind of tension (ex: a 7 Chord). Anyway, right now, in a 4/4 song, there is no way to render a measure shorter than 4 beats.

Oh ok got it. I would rate this as a nice to have since you have already a solution with the current functionality:

`2/4 G D/F# 4/4 Em` for example

Of course it could be done more elegantly but again, I think it is a nice to have and other enhancements would have priority over this one.

> In the end, the same result is accomplished. The only remaining "mystery" is how the measure bars will align? (ex: when displaying only chords)

For me this is straightforward, the rendering should not take into account the split of the measure because it only makes sense for lyrics. So if you are not displaying them, the rendering should  be as it is today

> Chord line where a measure should be split up (seems more natural to me this way).
```
|A                  |D...    \ 
_ Lorem ipsum dolor _sit amet,
G.   |A             |
_Con _sectetur adipiscing elit
```
Gut feeling is that the parsing will be easier to implement if the split is done at the lyrics level. I agree it makes the writing a bit more painful as you would probably get the lyrics with the proper formatting somewhere and then have to edit them to match the bars. And since writability is a core concern of ChordMark I think you have a point there.

So in the end:

```
A D... \ 
_ Lorem ipsum dolor _sit amet, 
G. A
_Con _sectetur adipiscing elit
```

I like this idea!

**Reply by thecodeinventor** [reply]
> 2/4 G D/F# 4/4 Em for example

This is the way I've been trying to handle such situations; it doesn't always work because "2/4" is the smallest measure acceptable but it's longer than the pick-up measure (which are typically 1 beat). 

What if "1/4" or any other "1/..." would become acceptable measures? (right now 1/4 does not render)  after all... it's 1 beat and it could be used (by exception) for cases just like these. I think/hope it wouldn't be difficult to include in the acceptable measures list. 
what do you say? :)

**Reply by no-chris** [reply]
That's an easy fix yes

**Reply by no-chris** [reply]
Should splitting a chordline be allowed inside a sub-beat group?

```
A D.. [E \ 
_ Lorem ipsum dolor _sit amet, 
A] G
_Con _sectetur adipiscing elit
```

This is probably edge case enough to be considered at a later stage wdyt?

**Reply by no-chris** [reply]
> This is the way I've been trying to handle such situations; it doesn't always work because "2/4" is the smallest measure acceptable but it's longer than the pick-up measure (which are typically 1 beat).

This will be fixed in the next release #666

**Reply by thecodeinventor** [reply]
> Should splitting a chordline be allowed inside a sub-beat group?

It would be nice to have (especially because it would not generate a special case where splitting is not permitted) but I think such cases are rare too (Tried to think of an example of such a use case, but can't find one). I know functionality comes at the cost of effort. So, if it's easy enough then there's no time like the present to implement; else: implement later :)   (P.S. I will be sure to let you know if I encounter examples where this is absolutely necessary)



> This will be fixed in the next release #666

Awesome! Looking forward for it! 💯 
I assume 1/4 will work just like any other time signature, but just to be sure: will subdivisions be possible in a 1/4 measure? ex: 1/4 [NC A] 4/4 E

**Reply by no-chris** [reply]
> I assume 1/4 will work just like any other time signature, but just to be sure: will subdivisions be possible in a 1/4 measure? ex: 1/4 [NC A] 4/4 E

works, added a test case for it https://github.com/no-chris/chord-mark/commit/acc576e3aff735cfc439acf3b2dfe29e6103111c

**Comment by thecodeinventor** [comment]
Once again, great job with this library! 🥇 Can't wait for the next update!

---

I have some more feedback for you:
in 6/8 time, rendering is weird:
6/8
`C. F. G. C. F. G.  ---> |C  F      |G        C|F        G |`
`C.. F.. G.. C..    ---> |C     |F   |G   |C   |`
`C. F.              ---> |C   F  |`
`C. F. G.           ---> error`
how do you write 3 chords in a 6/8 measure?

**Reply by no-chris** [reply]
You make me doubt
I always considered 6/8 as 2 beats with three 8th notes each
so with that definition ChordMark's rendering is I think consistent

It's also consistent with Wikipedia? https://en.wikipedia.org/wiki/Time_signature

**Reply by thecodeinventor** [reply]
I think you are right. I'm sorry I said it was "weird". My bad.

**Reply by no-chris** [reply]
No worries 😉
